File: Zilla_FPGA_Validation/SW/programs/and_op/and_op.py
#!/usr/bin/env python3
#
# This script generates header file for I-Type and U-Type tests
from sys import argv
import random

dest_file  = argv[1]               #dest file
#test_count = argv[2]               #test count

rnd_list = []

f= open(dest_file,"w+")

f.write("/*\r\n")
f.write("**\r\n")
f.write("*******************************************************************************\r\n")
f.write("**\r\n")
f.write("** Header file for creating I-Type and U-Type tests\r\n")
f.write("**\r\n")
f.write("** Usage : #include \"ui_and_test.h\"\r\n")
f.write("*******************************************************************************\r\n")
f.write("*/\r\n")
f.write("\r\n")

f.write("#ifndef UI_AND_TEST_H\r\n")
f.write("#define UI_AND_TEST_H\r\n")
f.write("\r\n")

f.write("#ifdef SPIKE_RUN\r\n")
f.write("#include \"data.h\"               /*!< Testbench created data */\r\n")
f.write("#endif\r\n")


#f.write("#define TEST_COUNT     (%d) /*!< Test count             */\r\n" % int(test_count))
f.write("\r\n")
f.write("\r\n")


f.write("int test_ui_and_type(int num)\r\n")
f.write("{\r\n")

f.write ("volatile int testnumber=num; \r\n")

f.write("#ifdef SPIKE_RUN\r\n")
f.write("	printf(\"\\n------------------------------\");\r\n")
f.write("	printf(\"\\nTest/> Instruction 8 bit: andi\");\r\n")
f.write("	printf(\"\\n------------------------------\\n\");\r\n")
f.write("#endif\r\n")

rnd_list.clear()
for i in range(15):
	rnd_list.append(random.randint(-2047,2048))


f.write("#ifdef SPIKE_RUN\r\n")
for i in range(15):

	f.write("{	TEST_ANDI_INSN_I_SPIKE8(andi, op1, %d, result, exp_result, testnumber);\r\n" % rnd_list[i])
	f.write("	print_test_results_andi_I8(testnumber, op1, op2, result, exp_result);} \r\n")
	f.write("	testnumber++;\r\n")
f.write("#else\r\n")
for i in range(15):
	f.write("{	TEST_ANDI_INSN_I8(MAILBOX1, andi, op1, %d, result, testnumber);}\r\n" % int(rnd_list[i]))
	f.write("	testnumber++;\r\n")
f.write("#endif\r\n")

f.write("#ifdef SPIKE_RUN\r\n")
f.write("	printf(\"\\n------------------------------\");\r\n")
f.write("	printf(\"\\nTest/> Instruction 16 bit: andi\");\r\n")
f.write("	printf(\"\\n------------------------------\\n\");\r\n")
f.write("#endif\r\n")

rnd_list.clear()
for i in range(15):
	rnd_list.append(random.randint(-2047,2048))


f.write("#ifdef SPIKE_RUN\r\n")
for i in range(15):

	f.write("{	TEST_ANDI_INSN_I_SPIKE16(andi, op1, %d, result, exp_result, testnumber);\r\n" % rnd_list[i])
	f.write("	print_test_results_andi_I16(testnumber, op1, op2, result, exp_result);} \r\n")
	f.write("	testnumber++;\r\n")
f.write("#else\r\n")
for i in range(15):
	f.write("{	TEST_ANDI_INSN_I16(MAILBOX1, andi, op1, %d, result, testnumber);}\r\n" % int(rnd_list[i]))
	f.write("	testnumber++;\r\n")
f.write("#endif\r\n")


f.write("#ifdef SPIKE_RUN\r\n")
f.write("	printf(\"\\n------------------------------\");\r\n")
f.write("	printf(\"\\nTest/> Instruction 32 bit: andi\");\r\n")
f.write("	printf(\"\\n------------------------------\\n\");\r\n")
f.write("#endif\r\n")

rnd_list.clear()
for i in range(15):
	rnd_list.append(random.randint(-2047,2048))


f.write("#ifdef SPIKE_RUN\r\n")
for i in range(15):

	f.write("{	TEST_ANDI_INSN_I_SPIKE32(andi, op1, %d, result, exp_result, testnumber);\r\n" % rnd_list[i])
	f.write("	print_test_results_andi_I32(testnumber, op1, op2, result, exp_result);} \r\n")
	f.write("	testnumber++;\r\n")
f.write("#else\r\n")
for i in range(15):
	f.write("{	TEST_ANDI_INSN_I32(MAILBOX1, andi, op1, %d, result, testnumber);}\r\n" % int(rnd_list[i]))
	f.write("	testnumber++;\r\n")
f.write("#endif\r\n")


f.write("#ifdef SPIKE_RUN\r\n")
f.write("	printf(\"\\n------------------------------\");\r\n")
f.write("	printf(\"\\nTest/> Instruction 64 bit: andi\");\r\n")
f.write("	printf(\"\\n------------------------------\\n\");\r\n")
f.write("#endif\r\n")

rnd_list.clear()
for i in range(15):
	rnd_list.append(random.randint(-2047,2048))

f.write("#ifdef SPIKE_RUN\r\n")
for i in range(15):

	f.write("{	TEST_ANDI_INSN_I_SPIKE64(andi, op1, %d, result, exp_result, testnumber);\r\n" % rnd_list[i])
	f.write("	print_test_results_andi_I64(testnumber, op1, op2, result, exp_result);} \r\n")
	f.write("	testnumber++;\r\n")
f.write("#else\r\n")
for i in range(15):
	f.write("{	TEST_ANDI_INSN_I64(MAILBOX1, andi, op1, %d, result, testnumber);}\r\n" % int(rnd_list[i]))
	f.write("	testnumber++;\r\n")
f.write("#endif\r\n")
f.write("\r\n")
f.write("\r\n")


f.write("return testnumber;")



f.write("}\r\n")
f.write("\r\n")

f.write("#endif /* UI_AND_TEST_H */\r\n")

f.close() 

File: Zilla_FPGA_Validation/SW/programs/and_op/and_op.c
/*
**
*******************************************************************************
**
** Test program for Logical AND
**
*******************************************************************************
*/

#include <stdint.h> 

#define TEST_COUNT     (5)    /*!< Test count */

int main(void)
{
   volatile int testnumber   = 1;
   uint32_t init[4]={0x123, 0x234, 0x345, 0x456}, length=4;
   
   init_by_array(init, length);

/*********************************** And Test uint8_t ********************************************/


	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
#ifdef SPIKE_RUN
uint8_t op1, op2; \
	                                  uint8_t rslt_full; \
					  uint8_t rslt_tb; \
                                          op1 = (uint8_t)data[(tn-1)*3 + 0]; \
                                          op2 = (uint8_t)data[(tn-1)*3 + 1]; \
                                          rslt_tb = (uint8_t)data[(tn-1)*3 + 2]; \
                                          rslt_full = op1 & op2;
           if(rslt_full == rslt_tb) 
	    printf("OP1 = %lu   OP2 = %lu F-result = %lu, S-result = %lu : Status = [PASS]\n\r", op1, op2, rslt_tb, rslt_full); 
	else 
	    printf("OP1 = %lu   OP2 = %lu F-result = %lu, S-result = %lu : Status = [FAIL]\n\r", op1, op2, rslt_tb, rslt_full);


#else 	
 		uint8_t op1, op2; 
		uint8_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res); 
		
#endif
	}
					

/**************************************************************************************************/

/*********************************** And Test uint16_t ********************************************/   


  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
#ifdef SPIKE_RUN






#else 
 		uint16_t op1, op2; 
		uint16_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res);  

#endif
	}

/**************************************************************************************************/

/*********************************** And Test uint32_t ********************************************/   


  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
 		uint32_t op1, op2; 
		uint32_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res); 	
    }

/**************************************************************************************************/ 

/*********************************** And Test 8 & 16 ********************************************/   


  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
 		uint8_t  op1;
        	uint16_t op2;
		uint16_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res); 	
    }

/**************************************************************************************************/

/*********************************** And Test 8 & 32 ********************************************/   


  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
 		uint8_t op1;
        	uint32_t op2; 
		uint32_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res); 	
    }

/**************************************************************************************************/ 

/*********************************** And Test 16 & 8 ********************************************/   


  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
 		uint16_t op1;
        	uint8_t  op2; 
		uint16_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res); 
    }

/**************************************************************************************************/ 

 /*********************************** And Test 16 & 32 ********************************************/   


  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
 		uint16_t op1;
        	uint32_t op2; 
		uint32_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res); 
    }

/**************************************************************************************************/ 

/*********************************** And Test 32 & 8 ********************************************/   


  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
 		uint32_t op1;
        	uint8_t  op2; 
		uint32_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res); 	
    }

/**************************************************************************************************/ 

/*********************************** And Test 32 & 16 ********************************************/   


  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
 		uint32_t op1;
        	uint16_t op2; 
		uint32_t res; 
		op1 = genrand_int32(); 
		op2 = genrand_int32(); 
		res = op1 & op2; 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res); 
    }

/**************************************************************************************************/

/*************************************** Assembly Test *************************************/

  	for(testnumber = 1; testnumber <= TEST_COUNT; testnumber++)
	{
#ifdef SPIKE_RUN

					uint8_t op1, op2; \
	                                  uint8_t rslt_full; \
                                          uint8_t rslt_tb; \
                                          op1 = (uint8_t)data[(tn-1)*3 + 0]; \
                                          op2 = (uint8_t)data[(tn-1)*3 + 1]; \
                                          rslt_tb = (uint8_t)data[(tn-1)*3 + 2]; \
                                          _AND_TEST_U8_(&op1, &op2, &rslt_full);




#else 
					  uint8_t op1, op2; 
	                                  uint8_t rslt_full; 
                                          op1 = genrand_int32(); 
                                          op2 = genrand_int32();                                           
				          _AND_TEST_U8_(&op1, &op2, &rslt_full); 
		printf_("OP1 = %lu   OP2 = %lu    Result = %lu\n\r",op1,op2,res);  

#endif
	}
					  uint8_t op1, op2; \
	                                  uint8_t rslt_full; \
                                          op1 = (uint8_t)generate_random_num(mb); \
                                          op2 = (uint8_t)generate_random_num(mb); \
				          _AND_TEST_U8_(&op1, &op2, &rslt_full); \
return 0;
}

